Sources/OktaAnalytics/OktaAnalytics.swift

```diff
@@ -111,21 +117,305 @@ public final class OktaAnalytics: NSObject {
     }
+
+    /**
+     removes all providers, scenarios from memory
+     */
+    public static func disposeAll() {
+        lock.writeLock()
+        defer { lock.unlock() }
+        disposeProviders()
+        disposeAllScenarios()
+    }
 }
 
+// Extensions
 public extension Dictionary {
     // Merge the contents of one dictionary into another, favoring the content of right
     static func mergeRecursive(left: inout Self, right: Self?) {
         left.merge(right ?? [:]) { current, _ in current }
     }
 }
 
+// Extensions
+extension OktaAnalytics {
+    /**
+        Starts an event scenario with the specified name.
+
+        - Parameters:
+           - eventName: The name of the event scenario to start.
+           - propertySubject: A closure that takes a `PassthroughSubject` of `Property` objects as a parameter.
+        */
+    public static func startScenario(_ scenario: ScenarioEvent, _ completion: @escaping (String) -> Void) {
+        lock.writeLock()
+        defer { lock.unlock() }
+        let properties = scenario.properties
+        let scenario = Scenario(id: UUID().uuidString, name: scenario.name, displayName: scenario.displayName, startTime: Date())
+
+        Self.providers.forEach {
+            $1.logger?.log(level: .debug, eventName: scenario.name, message: "\($0) Scenario \(scenario.name) in flight", properties: nil, file: #file, line: #line, funcName: #function)
+        }
+
+        if databasePool == nil {
+            Task(priority: .high) {
+                do {
+                    let databasePool = try await checkAndInitializeDB()
+                    self.databasePool = databasePool
+                    try await databasePool?.write {
+                        try scenario.insert($0)
+                    }
+                    updateScenario(scenario.id, properties)
+                    completion(scenario.id)
+                } catch {
+                    providers.forEach {
+                        $1.logger?.log(error: error as NSError, file: #file, line: #line, funcName: #function)
+                    }
+                }
+            }
+        } else {
+            do {
+                try databasePool?.write {
+                    try scenario.insert($0)
+                }
+                updateScenario(scenario.id, properties)
+                completion(scenario.id)
+            } catch {
+                providers.forEach {
+                    $1.logger?.log(error: error as NSError, file: #file, line: #line, funcName: #function)
+                }
+            }
+        }
+    }
+
+    /**
+        update a property scenario with the values.
+
+        - Parameters:
+           - scenarioID: unique sceario ID returned from `startScenario(_ , _)` .
+           - propertySubject: A closure that takes a `PassthroughSubject` of `Property` objects as a parameter.
+        */
+    public static func updateScenario(_ scenarioID: ScenarioID, _ properties: [Property]) {
+        lock.writeLock()
```

Comment:
I don't recommend such huge critical sections, not clear what exactly we are trying to synchronize